feat: server-enforced RpcInvokePermission and spoof prevention

xmanning opened the pull request:
## Purpose of this PR
When a server receives an RpcMessage, it should update the SenderClientId to match the SenderId provided by the messaging manager. This is to prevent modified clients from spoofing their SenderClientId as other clients.

### Changelog
- Fixed: Modified clients could spoof SenderClientId when sending an Rpc.

## Documentation
No documentation changes or additions were necessary.

## Testing & QA (How your changes can be verified during release Playtest)
Manual testing required.

xmanning committed:
Secure RpcMessage against SenderClientId spoofing

When a server receives an RpcMessage, it should update the SenderClientId to match the SenderId provided by the messaging manager. This is to prevent modified clients from spoofing their SenderClientId as other clients.

xmanning committed:
Merge branch 'develop-2.0.0' into develop-2.0.0

EmandM reviewed (commented):
Thank you very much for your contribution! We're aware from [this discussions thread](https://discussions.unity.com/t/only-allow-server-to-invoke-rpcs-on-clients/1685328) that there are maybe some more vulnerabilities in this area.

To ensure we get a thorough fix here, do you have any further thoughts on safety improvements or otherwise we could do in this area?

xmanning commented:
I'm not a network or cybersecurity engineer by any means so my thoughts may be a bit naïve, but I think server validation on Rpc calls are a good starting point. Aside from the server trusting the client to identify itself as this PR aims to fix, I believe [RequireOwnership is also enforced locally and can be easily bypassed. ](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/5f161817637a48cf23455362fd08a56d0fe3fb7f/com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs#L332) 
There are a few approaches to the issue in the discussion you linked that I think could fix that issue as well, such as a parameter on the RpcAttribute that replaces RequireOwnership, similar to the NetworkVariableWritePermission enum.
```cs
enum RpcInvokePermission
{
    Server,
    Owner,
    Everyone
}

// Usage
[Rpc(SendTo.Everyone, RpcInvokePermission.Server)] // Effectively the same as the legacy ClientRpcAttribute
private void DoSomeVfxRpc(...) { ... }

[Rpc(SendTo.Everyone, RpcInvokePermission.Owner)] // For the owner to notify clients of some event. Identical to RequireOwnership = true, but would be validated on the server.
private void PlayMonsterGrowlRpc(...) { ... }

[Rpc(SendTo.Everyone, RpcInvokePermission.Everyone)] // For things that don't need to be validated. Identical to the current implementation.
private void SendChatMessageRpc(...) { ... } 
```
Of course, this would need to be validated by the server when receiving Rpc messages.

Another option, although more complex, would be a user-defined validation middleware callback that returns a boolean to either allow or reject the Rpc, similar to how web frameworks like Express.js work. This approach would effectively be the same as the intended model of sending a Server Rpc that gets sent to clients after validation, but would allow for reusable middleware functions like "IsOwner", "IsServer", or to be more specific to certain use cases, something like "HasItem(ItemId)". 

Personally, I think an enum to control who can invoke an Rpc is the best solution, but there is no way around causing breaking changes.

I would like to note that from what I could tell, these issues are only within the modern RpcAttribute, and the legacy ServerRpc and ClientRpc do not have this issue. Acknowledging this vulnerability as just a risk of the system and recommending the legacy Rpc attributes for scenarios where security is important is always an option as well.

xmanning committed:
Replaced RequireOwnership with a new RpcInvokePermission
that is respected from the server for both direct and
proxy RPCs.

xmanning committed:
Replaced RequireOwnership with a new RpcInvokePermission
that is respected from the server for both direct and
proxy RPCs.

xmanning committed:
Add missing switch case

xmanning committed:
Merge branch 'develop-2.0.0' of https://github.com/xmanning/com.unity.netcode.gameobjects into develop-2.0.0

xmanning committed:
Another missing switch case

EmandM commented:
I really like the idea of the invoke permission. I think it's a nice and simple way to give a lot more flexibility and security to RPC calls.

Unfortunately, it's a breaking change for us to remove the `RequireOwnership` param. Seeing as it's an optional param we can change the order of the params, but we can't break any existing code.

If you open the testproject unity project inside this repository, you can see that this changeset as it exists breaks a number of existing tests.

We'd love to add this new functionality, is there any chance you can update these changes to handle both the `RequireOwnership` and the `InvokePermission` param? `InvokePermission` can override `RequireOwnership` and we can then mark `RequireOwnership` as obsolete.

xmanning committed:
Restore and deprecate RequireOwnership in favor of RpcInvokePermission

xmanning committed:
Only do permission validations on server

context.SenderId will be the server for a client receiving the proxied message, so the checks will not be accurate. The checks are also unnecessary considering the checks have already been done by the server.

EmandM commented:
I really like this approach! I will have time next week to see if we can land this functionality.

EmandM commented:
Hi, we've recently realized that our CLA check hasn't been running. I'm commenting here to trigger it to run.

EmandM committed:
Merge branch 'develop-2.0.0' into develop-2.0.0

xmanning committed:
ClientRpc properly assigned InvokePermission

Assigning ClientRpc.InvokePermission in the constructor was incorrect. Since we can check if it is a ClientRpc easily, I just do that in ILPP

xmanning committed:
Merge branch 'develop-2.0.0' of https://github.com/xmanning/com.unity.netcode.gameobjects into develop-2.0.0

xmanning committed:
Merge branch 'develop-2.0.0' into develop-2.0.0
